param/__init__.py

```diff
@@ -195,24 +185,34 @@ def guess_param_types(**kwargs):
         elif isinstance(v, tuple):
             if all(_is_number(el) for el in v):
                 params[k] = NumericTuple(**kws)
-            elif all(isinstance(el. dt_types) for el in v) and len(v)==2:
+            elif all(isinstance(el, dt_types) for el in v) and len(v)==2:
                 params[k] = DateRange(**kws)
             else:
                 params[k] = Tuple(**kws)
         elif isinstance(v, list):
             params[k] = List(**kws)
-        elif isinstance(v, np.ndarray):
-            params[k] = Array(**kws)
         else:
-            from pandas import DataFrame as pdDFrame
-            from pandas import Series as pdSeries
-
-            if isinstance(v, pdDFrame):
-                params[k] = DataFrame(**kws)
-            elif isinstance(v, pdSeries):
-                params[k] = Series(**kws)
+            try:
+                from numpy import ndarray
+            except ImportError:
+                pass
+            else:
+                if isinstance(v, ndarray):
+                    params[k] = Array(**kws)
+                    continue
+            try:
+                from pandas import DataFrame as pdDFrame
+                from pandas import Series as pdSeries
+            except ImportError:
+                pass
             else:
-                params[k] = Parameter(**kws)
+                if isinstance(v, pdDFrame):
+                    params[k] = DataFrame(**kws)
+                    continue
+                elif isinstance(v, pdSeries):
+                    params[k] = Series(**kws)
+                    continue
+            params[k] = Parameter(**kws)
```

Comment:
Thanks! I have trouble following the logic flow; is there any reason to use `else` here as opposed to putting that code directly into the `try` block after the import?

Reply:
I've removed the `else` blocks by putting the `isinstance` checks in the `try` blocks, as you suggested. The reason why I did it this way in the first place was that I usually try to have as less code as possible in `try` blocks. This made the code arguably ugly though!

The code is still a little ugly after the change, it's this way because I want to avoid as much as possible to have to import numpy or pandas if not necessary.

Reply:
I think it's slightly easier to follow now, though yes, definitely good to reduce `try` code where possible.